### PR TITLE
cabal-install: update curl transport to support Basic authentication

### PR DESCRIPTION
**Pull request by frasertweedale**
This PR updates the curl transport to support HTTP Basic authentication scheme.  There are a
couple of preliminary small refactors.  Best review the commits independently and in order.

Targeting branch `3.12` because I couldn't build `master` (GHC panic, don't worry about it).
It rebases cleanly onto `master`.

### QA Notes

Ensure that authenticated operations **using username and password** against `hackage.haskell.org` succeed, when using the Curl transport.

### Checklist

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [ ] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

**Comment by ulysses4ever** [comment]
I don't think we can merge it into 3.12. Please, target `master`.

**Comment by andreabedini** [comment]
@frasertweedale I rebased your PR on top of master. I hope it's ok.

**Comment by frasertweedale** [comment]
@andreabedini that's fine.  Note that we *do* want to backport this change for maintenance releases on older branches (you know the reasons).

If you prefer a variant of the patch for backports with the minimal changes and without the refactors, I can prepare that.  Or else, I suppose it lands in master first and then the backports can be created (is there automation for that?)

**Review by andreabedini — APPROVED** [review]
:+1:

File: cabal-install/src/Distribution/Client/HttpUtils.hs [hunk 1]
```diff
@@ -251,20 +251,35 @@ downloadURI transport verbosity uri path = do
...
+  transportCheckHttpsWithError verbosity transport uri $
+    TransportCheckHttps uri requiresHttpsErrorMessage
+
+-- | If the URI scheme is HTTPS, ensure the transport supports HTTPS.
+-- If not, fail with the given error.
+transportCheckHttpsWithError
+  :: Verbosity -> HttpTransport -> URI -> CabalInstallException -> IO ()
+transportCheckHttpsWithError verbosity transport uri err
+  | isHttpsURI uri
   , not (transportSupportsHttps transport) =
```
Comment (ffaf1) [inline cabal-install/src/Distribution/Client/HttpUtils.hs:273]: I understand this is not your code, but is that `,` equivalent to `&&`?

If so, why not use `&&`, which is clearer in its intention?

Comment (geekosaur) [reply cabal-install/src/Distribution/Client/HttpUtils.hs:273]: It is, per the Report. Which one is clearer depends on context.

**Review by ffaf1 — APPROVED** [review]
good afaics

**Comment by ulysses4ever** [comment]
@frasertweedale thank you for your contribution! As I mentioned above, we only accept non-master-targeting PRs as exceptions. Could you please create a PR against `master`? If you don't have time, we can look into doing it ourselves, but this may slow down it quite a bit.

**Comment by ulysses4ever** [comment]
Please, disregard my previous message, I was wrong about the branch.

Let me put a label on this so that it gets merged.

**Comment by andreasabel** [comment]
@frasertweedale wrote:
> * [x]  [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Where can I find those?

Did you test uploading documentation (`cabal upload -d ...`)?

**Pull request merged**